wstl/pswstl2milp.py

```diff
@@ -0,0 +1,327 @@
+             that needs to be satisfied at the specific time. Note that Disjunction
+             and eventually are special cases'''    
+        ret = set()
+        if formula.op == Operation.PRED:
+            if self.variables[formula][t].x == 1:
+                ret = {(formula, t)}
+            else:
+                ret = {}
+
+        elif formula.op == Operation.AND:
+            for f in formula.children:
+                ret = ret.union(self.predicate_pairs(f, t))
+
+        elif formula.op == Operation.OR:
+            for f in formula.children:
+                if self.variables[f][t].x == 1:
+                    ret = ret.union(self.predicate_pairs(f, t))
+                    break
+
+        elif formula.op == Operation.ALWAYS:
+            f = formula.child
+            interval = range(int(formula.low), int(formula.high+1))
+            for t in interval:
+                ret = ret.union(self.predicate_pairs(f, t))
+
+        elif formula.op == Operation.EVENT:
+            f = formula.child
+            interval = range (int(formula.low), int(formula.high+1))
+            for t in interval:
+                if self.variables[f][t].x == 1:
+                    ret = ret.union(self.predicate_pairs(f, t))
+                    break
+        return ret
+
+    def hierarchical(self, model_name='model_test.lp', optimize=True):
+        '''
+        This method computes a hierarchical optimization formulation 
+        (lexicografical) from root node all the way to the leaves (predicates)
+        Input:
+            - model_name is a file name to generate Gurobi information about 
+              the optimization problem
+            - optimize is a flag type variable which is True by default performing
+              the optimization of the problem, in case it is False it will only
+              generate the objective function.
+        
+        Output:
+            - depth of the formula
+        '''
+        max_depth = max(self.objectives)
+        for d in range(max_depth+1):
+            self.model.setObjectiveN(-self.objectives[d], d, 
+                                     priority=max_depth-d)
+            self.model.update()
+        
+        if optimize is True:
+            self.model.optimize()
+            self.model.write(model_name)
+        return d
+    
+    def ldf(self, model_name='model_test.lp', optimize=True):
+        '''
+        This method computes a Lowest depht first optimization formulation 
+        making and increasing penalization from being far from the root node
+        Input:
+            - model_name is a file name to generate Gurobi information about 
+              the optimization problem
+            - optimize is a flag type variable which is True by default performing
+              the optimization of the problem, in case it is False it will only
+              generate the objective function.
+        ''' 
+        M2 = 20 # FIXME: computed based on formula size
```

Comment:
Add `M2` as a parameter of the method with a default value. Also, add it as a list item to the `TODO.md`.

Reply:
Give the method argument a representative name.